Fix CUVID crash on resolution change

j0sh opened the pull request:
Also adds a bunch of other changes necessary to better support
mid-stream resolution changes.

Unfortunately with CUVID there still seems to be a brief flash of
green (looks to be the length of the decoder's internal frame buffer)
but we can tackle that separately. This PR simply makes the transcoder
1) not crash, and
2) correctly encode mid-stream rotations, including with CPUs

Depends on https://github.com/livepeer/lpms/pull/417

The following commits can all stand alone except the last one (unit tests) depends on all of them. I can make separate PRs if that is preferable for ease of review. Let's also keep these commits separate when merging.

[ffmpeg: Flush filters before re-initialization.](https://github.com/livepeer/lpms/commit/dd12875e9b5ce2114b6aa1be2ae1a35f764cb296) 
Also add another condition for re-initialization: if the
input resolution changes. This triggers the filter graph
to re-build and adjust to the new resolution, when CPU
encoders are in use. Resolves a long-standing TODO.

[ffmpeg: Re-init encoder on resolution change.](https://github.com/livepeer/lpms/commit/133d1832e977d87ea63d8351a5c1e34d51c4cfd6)
Necessary to have the encoder actually pick up the new
resolution correctly.

[ffmpeg: Reset the flush packet after each keyframe.](https://github.com/livepeer/lpms/commit/26fdf8cb1af355565b9e81bea6bc9c25b6fa0587) 
This handles cases where the packet may contain a frame
that triggers a decoder reset - we do not want to cause
a reset during the flushing process.

[ffmpeg: Handle EAGAIN from decoder and drain](https://github.com/livepeer/lpms/commit/6db145ea5a96d11dd1b27200f6990ab74f3f0d18) 
This usually happens with CUVID if the decoder needs to be reset
internally for whatever reason, such as a mid-stream resolution
change. Previously, EAGAIN would loop and cause a crash.

Also block demuxing until decoder is ready to receive packets again.

[ffmpeg: Add tests for rotation.](https://github.com/livepeer/lpms/commit/986324ce939c99298c736d59f277a084bfadb0c6)

mjh1 commented on ffmpeg/encoder.c:
```diff
@@ -332,12 +336,83 @@ int reopen_output(struct output_ctx *octx, struct input_ctx *ictx)
+            pkt->pts = (int64_t)pkt->opaque; // already in filter timebase
+            pkt->dts = pkt->pts - av_rescale_q(pts_dts, encoder->time_base, time_base);
+          }
+          //fprintf(stderr, "JOSH - flushing %"PRId64"/%"PRId64"\n", pkt->dts, pkt->pts);
```
probably want to remove these commented out lines

j0sh replied:
good catch, thanks - fixed in 4175186fe5718b6c0ed0cc655c1856bfd914c707

mjh1 commented on ffmpeg/ffmpeg_test.go:
```diff
@@ -1927,3 +1927,248 @@ PTS_EOF
+	assert.Equal(t, 181, res.Encoded[0].Frames) // should be 180 ... ts rounding ?
+	assert.Equal(t, 360, res.Encoded[1].Frames)
+
+	// TODO test rollover of gop interval during flush
```
I'm not sure what we normally do in this repo for TODOs but you maybe want to raise tickets instead if they're important to come back to so we don't forget.

j0sh replied:
this specific note is not so important, it is more noting a gap in the test coverage

the other TODOs sprinkled around these tests relate to discrepancies between cpu/gpu behavior are likely related to the green screen flash which we already have ticket(s) for

mjh1 reviewed (approved):
Looks good but I'm afraid I don't really know this codebase and C sorry

j0sh commented:
Squashed the comments from the PR review, changed the merge target to `master` so we will also pull in #417, then rebased onto latest master (I had been working off an older tip which had merge conflicts)

The CI-driven unit tests will probably fail right now even though they pass locally; see https://github.com/livepeer/lpms/pull/415#issuecomment-2276392076
